feat: [HELP WANTED] Adding custom memory budget for search index

vaibhawvipul opened the pull request:
# Ticket(s) Closed

- Closes # #1505 

## What
- custom memory budget for tantivy search indexes
- default is 500_000_000
- custom cache block size for readers
- default is 100

## Why
- gives more fine grained control to user
- helps in performance tuning

## How
- user can create index with an 'memory_budget' and 'reader_cache_num_blocks' option.

## Tests

philippemnoel reviewed (commented):
This looks clean!! A few questions:

1) Could you please write documentation for this? You can inspire yourself from #1506 
2) See the issue, what about Reader Cache and Segment Configuration?

neilyio commented on pg_search/src/index/search.rs:
```diff
@@ -213,9 +217,14 @@ impl SearchIndex {
-            .underlying_index
-            .writer(INDEX_TANTIVY_MEMORY_BUDGET)?;
+
+        let memory_budget = if search_index.memory_budget < 15_000_000 {
+            INDEX_TANTIVY_MEMORY_BUDGET_DEFAULT
+        } else {
+            search_index.memory_budget
+        };
```
I think we should make the `15_000_000` a const and make sure we have doc comments to explain the number.

neilyio commented on pg_search/src/index/search.rs:
```diff
@@ -314,42 +323,46 @@ impl<'de> Deserialize<'de> for SearchIndex {
     where
         D: Deserializer<'de>,
     {
-        // A helper struct that lets us use the default serialization for most fields.
         #[derive(Deserialize)]
         struct SearchIndexHelper {
             schema: SearchIndexSchema,
             directory: WriterDirectory,
-            // An index created in an older version of pg_search may not have serialized a uuid
-            // to disk. Just use an empty string for backwards compatibility.
             #[serde(default)]
             uuid: String,
```
Why remove these comments?

vaibhawvipul replied:
great catch. this was removed mistakenly. restored.

neilyio commented on pg_search/src/postgres/options.rs:
```diff
@@ -125,6 +126,16 @@ extern "C" fn validate_uuid(value: *const std::os::raw::c_char) {
     cstr_to_rust_str(value);
 }
 
+#[pg_guard]
+extern "C" fn validate_memory_budget(value: *const std::os::raw::c_char) {
+    let memory_budget_str = cstr_to_rust_str(value);
+    if !memory_budget_str.is_empty() {
+        memory_budget_str
+            .parse::<i32>()
+            .expect("Invalid memory budget value");
+    }
+}
+
```
If the user is going to see this error message, we should try and make it as helpful as possible so they can be aware of the valid range of values.

vaibhawvipul commented:
> This looks clean!! A few questions:
> 
> 1. Could you please write documentation for this? You can inspire yourself from [docs: Document performance tuning #1506](https://github.com/paradedb/paradedb/pull/1506)
> 2. See the issue, what about Reader Cache and Segment Configuration?

@philippemnoel 
1. I have added documentation, raise a PR, please take a look - https://github.com/paradedb/paradedb/pull/1532
2. I have added Reader Cache. I would request to do Segment configuration in a follow-up issue. Let us keep this only for memory related changes. If that is fine with you.

philippemnoel reviewed (approved):
High level docs are in: https://github.com/paradedb/paradedb/pull/1506/files

Perhaps we should document how the user can configure this setting?

philippemnoel commented:
Closing this until we're ready to explore it again
